src/custom_links/singly_linked_list.rs

```diff
@@ -0,0 +1,1485 @@
+// Copyright 2020 Amari Robinson
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use core::cell::Cell;
+use core::fmt;
+
+use super::link_ops::{self, DefaultLinkOps};
+use super::Adapter;
+use super::pointer_ops::PointerOps;
+
+// =============================================================================
+// SinglyLinkedListOps
+// =============================================================================
+
+/// Link operations for `SinglyLinkedList`.
+pub unsafe trait SinglyLinkedListOps: super::LinkOps {
+    fn next(&self, ptr: Self::LinkPtr) -> Option<Self::LinkPtr>;
+
+    unsafe fn set_next(&mut self, ptr: Self::LinkPtr, next: Option<Self::LinkPtr>);
+}
+
+// =============================================================================
+// Link
+// =============================================================================
+
+/// Intrusive link that allows an object to be inserted into a
+/// `SinglyLinkedList`.
+pub struct Link {
+    next: Cell<*const Link>,
```

Comment:
Change this to `Cell<Option<NonNull<Link>>>`. Then `LinkPtr` can just be `NonNull<Link>`.

Reply:
The what's the new value of `UNLINKED_MARKER`?
```rust
const UNLINKED_MARKER: Option<NonNull<Link>> = unsafe { Some(NonNull::new_unchecked(1 as *mut Link)) };
// or
const UNLINKED_MARKER: Option<NonNull<Link>> = None;
```

Reply:
The first one.